enterprise/coderd/roles.go

```diff
@@ -12,70 +14,86 @@ import (
 	"github.com/coder/coder/v2/codersdk"
 )
 
-// patchRole will allow creating a custom role
-//
-// @Summary Upsert a custom site-wide role
-// @ID upsert-a-custom-site-wide-role
-// @Security CoderSessionToken
-// @Produce json
-// @Tags Members
-// @Success 200 {array} codersdk.Role
-// @Router /users/roles [patch]
-func (api *API) patchRole(rw http.ResponseWriter, r *http.Request) {
-	ctx := r.Context()
+type enterpriseCustomRoleHandler struct {
+	Enabled bool
+}
 
-	var req codersdk.Role
-	if !httpapi.Read(ctx, rw, r, &req) {
-		return
+func (h enterpriseCustomRoleHandler) PatchOrganizationRole(ctx context.Context, db database.Store, rw http.ResponseWriter, orgID uuid.UUID, role codersdk.Role) (codersdk.Role, bool) {
+	if !h.Enabled {
+		httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+			Message: "Custom roles is not enabled",
+		})
+		return codersdk.Role{}, false
 	}
 
-	if err := httpapi.NameValid(req.Name); err != nil {
+	// Only organization permissions are allowed to be granted
+	if len(role.SitePermissions) > 0 {
 		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-			Message: "Invalid role name",
-			Detail:  err.Error(),
+			Message: "Invalid request, not allowed to assign site wide permissions for an organization role.",
+			Detail:  "organization scoped roles may not contain site wide permissions",
 		})
-		return
+		return codersdk.Role{}, false
 	}
 
-	if len(req.OrganizationPermissions) > 0 {
-		// Org perms should be assigned only in org specific roles. Otherwise,
-		// it gets complicated to keep track of who can do what.
+	if len(role.UserPermissions) > 0 {
 		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-			Message: "Invalid request, not allowed to assign organization permissions for a site wide role.",
-			Detail:  "site wide roles may not contain organization specific permissions",
+			Message: "Invalid request, not allowed to assign user permissions for an organization role.",
+			Detail:  "organization scoped roles may not contain user permissions",
 		})
-		return
+		return codersdk.Role{}, false
+	}
+
+	if len(role.OrganizationPermissions) > 1 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid request, Only 1 organization can be assigned permissions",
+			Detail:  "roles can only contain 1 organization",
+		})
+		return codersdk.Role{}, false
+	}
```

Comment:
If this is an invalid state, why is it representable in a `codersdk.Role`?

Reply:
It silently omits permissions. Which feels a bit off.

https://github.com/coder/coder/blob/553dca296bd0970cfab11a2f5e82ab734b6ed3a2/coderd/database/db2sdk/db2sdk.go#L540-L542

Returning an error feels like it could have a single role "break" things. Wondering if I could include an extra field with like `warnings` :thinking:

Reply:
Returning an error doesn't have to specifically break things, we could export `ErrNoMultiOrgRole` and `IsNoMultiOrgRoleError()` in db2sdk and handle them appropriately by dropping an error log. This would at least allow us to detect this in tests.

However, it feels like the 'right' fix here is to just not allow multi-org roles at all in `rbac`. It's not a blocker to this PR, but it feels like something we should fix.

Reply:
Yea, it might be worth refactoring the rbac to just prevent this altogether :thinking:. I think it could be done without trickling down to the rego.

I think that is the better approach, as I can't see a reason for it in the future.